### ISSUE TITLE
Feature help

### ISSUE THREAD

**Issue by febg11** [linkedin/LiTr#34]
Hi,

I have used ffmpeg before to compress a video to a specific file size e.g. 20MB by using 2 pass encoding and calculating the bitrate to use, an example can be seen in this link https://trac.ffmpeg.org/wiki/Encode/H.264. 

However I can't really use ffmpeg because of the GPL license and MPEG LA patents. 

Would I be able to achieve something similar using this library? I haven't really done much with the media codec or anything this low level but I would appreciate if you could give me some guidance. 

Thanks a lot

**Comment by izzytwosheds** [comment]
So that depends on how exact you want that size to be. LiTr relies on MediaCodec, which usually relies on encoder hardware on a device. These codecs are optimized for camera recording, so they prioritize speed over accuracy. Because of that, they don't do 2-pass encoding or CBR (constant bitrate) encoding. Using VBR codec provides decently accurate size estimation (bitrate * duration), but that will not be precise. 

You don't need to know MediaCodec to use LiTr, it abstract all that stuff for you. LiTr comes with a very usable sample app that will let you do transcoding with different resolution/bitrate, it should help you play around with different settings and decide if LiTr does what you need it to do. Feel free to ask more questions, I will keep this issue open.

**Comment by febg11** [comment]
Thanks for replying. I'm actually creating a flutter app so I have to create a wrapper to call this in android. I have looked at your sample app but I am not great with android. Occassionally my code runs but most of the time I am getting this error. i have made sure to allow storage settings. 
```

W/1.raster(11173): type=1400 audit(0.0:8633592): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=18003 scontext=u:r:untrusted_app:s0:c127,c258,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
W/VideoCapabilities(11173): Unsupported mime image/vnd.android.heic
W/VideoCapabilities(11173): Unrecognized profile 2130706433 for video/avc
W/VideoCapabilities(11173): Unrecognized profile 2130706434 for video/avc
W/VideoCapabilities(11173): Unrecognized profile 2130706433 for video/avc
W/VideoCapabilities(11173): Unrecognized profile 2130706434 for video/avc
W/VideoCapabilities(11173): Unsupported mime video/divx
W/VideoCapabilities(11173): Unsupported mime video/divx4
W/VideoCapabilities(11173): Unrecognized profile 4 for video/hevc
W/VideoCapabilities(11173): Unrecognized profile/level 0/3 for video/mpeg2
W/VideoCapabilities(11173): Unrecognized profile/level 0/3 for video/mpeg2
W/VideoCapabilities(11173): Unsupported mime video/x-ms-wmv
W/VideoCapabilities(11173): Unrecognized profile 2130706433 for video/avc
W/VideoCapabilities(11173): Unrecognized profile 2130706434 for video/avc
I/VideoCapabilities(11173): Unsupported profile 4 for video/mp4v-es
I/MediaCodec(11173): name: OMX.qcom.video.encoder.avc
E/MediaCodec(11173): MediaCodec::CreateByComponentName:OMX.qcom.video.encoder.avc
I/OMXClient(11173): IOmx service obtained
I/MediaCodec(11173): configure format is :{color-format=2130708361, durationUs=22645022, mime=video/avc, width=1280, bitrate=5000000, frame-rate=30, height=720}
W/MediaCodec(11173): RETYR:0
I/ExtendedACodec(11173): setupVideoEncoder()
W/OMXUtils(11173): do not know color format 0x7fa30c06 = 2141391878
W/OMXUtils(11173): do not know color format 0x7fa30c04 = 2141391876
W/OMXUtils(11173): do not know color format 0x7fa30c00 = 2141391872
W/OMXUtils(11173): do not know color format 0x7fa30c09 = 2141391881
W/OMXUtils(11173): do not know color format 0x7fa30c0a = 2141391882
W/OMXUtils(11173): do not know color format 0x7fa30c08 = 2141391880
W/OMXUtils(11173): do not know color format 0x7fa30c07 = 2141391879
W/OMXUtils(11173): do not know color format 0x7f000789 = 2130708361
E/ACodec  (11173): [OMX.qcom.video.encoder.avc] configureCodec returning error -38
E/ACodec  (11173): signalError(omxError 0x80001001, internalError -2147483648)
E/MediaCodec(11173): Codec reported err 0x80001001, actionCode 0, while in state 3
E/MediaCodec(11173): configure failed with err 0x80001001, resetting...,sleep 0.5s
E/MediaCodec(11173): configure failed with err 0x80001001, resetting...,sleep 0.5s done
I/OMXClient(11173): IOmx service obtained
E/com.linkedin.android.litr.exception.TrackTranscoderException(11173): Failed to retrieve media codec info.
E/TransformationJob(11173): Transformation job error
E/TransformationJob(11173): com.linkedin.android.litr.exception.TrackTranscoderException: Failed to configure encoder codec.Media transformation failed for job id: null
E/TransformationJob(11173): Media format: {color-format=2130708361, durationUs=22645022, mime=video/avc, width=1280, bitrate=5000000, frame-rate=30, height=720}
E/TransformationJob(11173): Selected media codec info: 
E/TransformationJob(11173): Available media codec info list (Name, IsEncoder, Supported Types): 
E/TransformationJob(11173): MediaCodecInfo: OMX.google.aac.decoder,false,[audio/mp4a-latm]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.amrnb.decoder,false,[audio/3gpp]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.amrwb.decoder,false,[audio/amr-wb]
E/TransformationJob(11173): MediaCodecInfo: OMX.qti.audio.decoder.flac,false,[audio/flac]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.g711.alaw.decoder,false,[audio/g711-alaw]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.g711.mlaw.decoder,false,[audio/g711-mlaw]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.gsm.decoder,false,[audio/gsm]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.mp3.decoder,false,[audio/mpeg]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.opus.decoder,false,[audio/opus]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.raw.decoder,false,[audio/raw]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.vorbis.decoder,false,[audio/vorbis]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.aac.encoder,true,[audio/mp4a-latm]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.amrnb.encoder,true,[audio/3gpp]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.amrwb.encoder,true,[audio/amr-wb]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.flac.encoder,true,[audio/flac]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.encoder.heic,true,[image/vnd.android.heic]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.avc,false,[video/avc]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.avc.secure,false,[video/avc]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.h264.decoder,false,[video/avc]
E/TransformationJob(11173): MediaCodecInfo: OMX.qti.video.decoder.divxsw,false,[video/divx]
E/TransformationJob(11173): MediaCodecInfo: OMX.qti.video.decoder.divx4sw,false,[video/divx4]
E/TransformationJob(11173): MediaCodecInfo: OMX.qti.video.decoder.h263sw,false,[video/3gpp]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.h263.decoder,false,[video/3gpp]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.hevc,false,[video/hevc]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.hevc.secure,false,[video/hevc]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.hevc.decoder,false,[video/hevc]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.mpeg2,false,[video/mpeg2]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.mpeg2.secure,false,[video/mpeg2]
E/TransformationJob(11173): MediaCodecInfo: OMX.qti.video.decoder.mpeg4sw,false,[video/mp4v-es]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.mpeg4.decoder,false,[video/mp4v-es]
E/TransformationJob(11173): MediaCodecInfo: OMX.qti.video.decoder.vc1sw,false,[video/x-ms-wmv]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.vp8,false,[video/x-vnd.on2.vp8]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.vp8.decoder,false,[video/x-vnd.on2.vp8]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.vp9,false,[video/x-vnd.on2.vp9]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.decoder.vp9.secure,false,[video/x-vnd.on2.vp9]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.vp9.decoder,false,[video/x-vnd.on2.vp9]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.encoder.avc,true,[video/avc]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.h264.encoder,true,[video/avc]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.encoder.h263sw,true,[video/3gpp]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.h263.encoder,true,[video/3gpp]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.encoder.hevc,true,[video/hevc]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.encoder.mpeg4sw,true,[video/mp4v-es]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.mpeg4.encoder,true,[video/mp4v-es]
E/TransformationJob(11173): MediaCodecInfo: OMX.qcom.video.encoder.vp8,true,[video/x-vnd.on2.vp8]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.vp8.encoder,true,[video/x-vnd.on2.vp8]
E/TransformationJob(11173): MediaCodecInfo: OMX.google.vp9.encoder,true,[video/x-vnd.on2.vp9]Diagnostic info: android.media.MediaCodec.error_neg_2147479551
E/TransformationJob(11173): 	at com.linkedin.android.litr.codec.MediaCodecEncoder.init(MediaCodecEncoder.java:70)
E/TransformationJob(11173): 	at com.linkedin.android.litr.transcoder.VideoTrackTranscoder.initCodecs(VideoTrackTranscoder.java:74)
E/TransformationJob(11173): 	at com.linkedin.android.litr.transcoder.VideoTrackTranscoder.<init>(VideoTrackTranscoder.java:64)
E/TransformationJob(11173): 	at com.linkedin.android.litr.transcoder.TrackTranscoderFactory.create(TrackTranscoderFactory.java:78)
E/TransformationJob(11173): 	at com.linkedin.android.litr.TransformationJob.createTrackTranscoders(TransformationJob.java:171)
E/TransformationJob(11173): 	at com.linkedin.android.litr.TransformationJob.transform(TransformationJob.java:94)
E/TransformationJob(11173): 	at com.linkedin.android.litr.TransformationJob.run(TransformationJob.java:74)
E/TransformationJob(11173): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:458)
E/TransformationJob(11173): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/TransformationJob(11173): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
E/TransformationJob(11173): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
E/TransformationJob(11173): 	at java.lang.Thread.run(Thread.java:764)
E/TransformationJob(11173): Caused by: android.media.MediaCodec$CodecException: Error 0x80001001
E/TransformationJob(11173): 	at android.media.MediaCodec.native_configure(Native Method)
E/TransformationJob(11173): 	at android.media.MediaCodec.configure(MediaCodec.java:1965)
E/TransformationJob(11173): 	at android.media.MediaCodec.configure(MediaCodec.java:1892)
E/TransformationJob(11173): 	at com.linkedin.android.litr.codec.MediaCodecEncoder.init(MediaCodecEncoder.java:54)
E/TransformationJob(11173): 	... 11 more
```

I am calling my the transform like so 
```
    public void tryCompressVideo(final Result result, final String path, final String outputPath) {

        MediaFormat targetVideoFormat = new MediaFormat();
        targetVideoFormat.setString(MediaFormat.KEY_MIME, "video/avc");
        targetVideoFormat.setInteger(MediaFormat.KEY_WIDTH, 1280);
        targetVideoFormat.setInteger(MediaFormat.KEY_HEIGHT, 720);
        targetVideoFormat.setInteger(MediaFormat.KEY_BIT_RATE, 5000000);
        // targetVideoFormat.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, 5);
        targetVideoFormat.setInteger(MediaFormat.KEY_FRAME_RATE, 30);

        String id = UUID.randomUUID().toString();
        final MediaTransformer mediaTransformer = new MediaTransformer(applicationContext);
        mediaTransformer.transform(id,
                Uri.parse(path),
                outputPath,
                targetVideoFormat,
                null,
                new TransformationListener() {

                    @Override
                    public void onStarted(@NonNull String id) {
                        int x = 1;
                        System.out.println(id);
                    }

                    @Override
                    public void onProgress(@NonNull String id, float progress) {
                        int x = 1;
                        System.out.println(id);
                        System.out.println(progress);

                    }

                    @Override
                    public void onCompleted(@NonNull String id, @Nullable List<TrackTransformationInfo> trackTransformationInfos) {
                        int x = 1;
                        System.out.println(id);
                        result.success(outputPath);
                    }

                    @Override
                    public void onCancelled(@NonNull String id, @Nullable List<TrackTransformationInfo> trackTransformationInfos) {
                        int x = 1;
                        System.out.println(id);

                    }

                    @Override
                    public void onError(@NonNull String id, @Nullable Throwable cause, @Nullable List<TrackTransformationInfo> trackTransformationInfos) {
                        int x = 1;
                        System.out.println(id);

                    }
                },
                MediaTransformer.GRANULARITY_DEFAULT,
                null);


    }
```
Could you help me understand what I am doing wrong?

**Comment by izzytwosheds** [comment]
Hmm, looks right to me overall. Try adding this:
targetVideoFormat.setInteger(MediaFormat.KEY_COLOR_FORMAT, MediaCodecInfo.CodecCapabilities.COLOR_FormatSurface);
And uncomment targetVideoFormat.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, 5);

Did you try compiling and running LiTr demo app for the same video file with the same parameters? Does it work?

**Comment by izzytwosheds** [comment]
It does look like that you have an encoder codec for AVC (hardware accelerated, by Qualcomm), but it is failing to configure correctly. Which device is this running on?

**Comment by febg11** [comment]
It's running on a oneplus 6

**Comment by febg11** [comment]
I just tried running the video in demo app with the Transcode video\audio (GL renderer) option and it does work fine.

**Comment by izzytwosheds** [comment]
That is good news. That means the reason is a programming error which is fixable. 
I tried commenting out setting key frame interval and got the same error as you. This confirms that encoder cannot initialize without that parameter. Try uncommenting it. If your UI doesn't yet support that, set it to 5. (value really doesn't matter all that much)

**Comment by febg11** [comment]
Ah that fixed it, sorry for asking a question for such a simple problem. I have managed to get the compression to a target file size by adjusting the bit rate working, your library is really good! 

I have a few more small questions  if you don't mind answering them

1) I have set the audio bitrate (see code below) when i encode the video. Similar to not including the KEY_I_FRAME_INTERVAL in the video i noticed that if I missed one of those 3 audio settings it would just crash. Are these crashes based on the device or is setting these parameters just compulsory. I am worried it may run on my phone and break on other brands because I haven't set a certain MediaFormat params. I have this concern for the video settings as well. 
```
final int AUDIO_SAMPlE_RATE = 48000;
final int AUDIO_CHANNELS = 2;
final int AUDIO_BITRATE = 128000;
... 
...
...
MediaFormat audioVideoFormat = new MediaFormat();
audioVideoFormat.setString(MediaFormat.KEY_MIME, MediaFormat.MIMETYPE_AUDIO_AAC);
audioVideoFormat.setInteger(MediaFormat.KEY_BIT_RATE, AUDIO_BITRATE);
audioVideoFormat.setInteger(MediaFormat.KEY_SAMPLE_RATE, AUDIO_SAMPlE_RATE);
audioVideoFormat.setInteger(MediaFormat.KEY_CHANNEL_COUNT, AUDIO_CHANNELS);
```

2) I just want to double check that it is ok to use this in a commercial application and that none of this code will be calling MPEG LA parented code. 

3) Is it important to be calling mediaTransformer.release(); in the complete, cancel and error callbacks?

Thanks for your help :)

**Comment by izzytwosheds** [comment]
I am happy LiTr is working out for your app!

1. Those are three things that must be defined for any audio stream for any compression. I think, this should be enough for any device. I have not experimented with audio transcoding a lot, though, and LiTr can only change audio bitrate. So make sure you are using the source audio's sampling rate and channel count values for target audio.
2. I have not dug deep on this. LiTr uses Android MediaCodec APIs, which many commercial apps use so I would assume it is ok. Still wouldn't hurt to double check with someone from legal who would know a lot more than us about specifics of this.
3. MediaTransformer lifecycle is up to you. If your app uses it throughout its lifecycle, it would make sense to create MediaTransformer on app/Activity start, and release it on its destruction. 
4. iOS does transcoding out of the box. You can start by looking at AVAssetExportSession: https://developer.apple.com/documentation/avfoundation/avassetexportsession I am not iOS media expert, I don't know iOS media stack well, I just looked at it at some point. I remember this class could do some basic preset based transcoding.

**Comment by izzytwosheds** [comment]
If you end up writing a nice Flutter wrapper for LiTr, it would be awesome if you open sourced it. I think it would very helpful to many devs that work with Flutter. 
Also, can I ask which app you work on? I would love to know which apps use LiTr and how (for bragging rights mostly)

**Comment by febg11** [comment]
Thanks for the answers :) I have switched my audio to use the source settings as you have suggested.

My app its just a personal project that i am working on in my spare time, not much to brag about ;) ... it was close to being finished until I realised ffmpeg was GPL and had to stop using it (this is why I wanted to double check so I don't make the same mistake twice). Using your library has really helped me with replacing the android side... Lets hope iOS goes as well.

I will look into the AVAssetExportSession and hopefully making an open source library that people can use... Although I haven't written any swift in my life so who knows what will happen.

Thanks again for the library and all the help.

**Comment by izzytwosheds** [comment]
Speaking of swift, sounds like this library might make your life easier: https://github.com/NextLevel/NextLevelSessionExporter

**Comment by febg11** [comment]
Hi again :) 
I have made some progress with the iOS library you suggested... It works similar to Litr which is nice.

I have  another question regarding Litr. I get tons of errors in my run output log... seems to just be when the app starts but I am not sure.  I kinda glazed over them while i was writing the code as they didn't seem to break anything but now they are bothering me. 

The errors are like this 
```
W/Thread-8( 5453): type=1400 audit(0.0:13061588): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=18003 scontext=u:r:untrusted_app:s0:c156,c258,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
E/libc    ( 5453): Access denied finding property "vendor.debug.egl.swapinterval"
W/Thread-8( 5453): type=1400 audit(0.0:13061589): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=18003 scontext=u:r:untrusted_app:s0:c156,c258,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
W/Thread-8( 5453): type=1400 audit(0.0:13061590): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=18003 scontext=u:r:untrusted_app:s0:c156,c258,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
E/libc    ( 5453): Access denied finding property "vendor.debug.egl.swapinterval"
E/libc    ( 5453): Access denied finding property "vendor.debug.egl.swapinterval"
W/Thread-8( 5453): type=1400 audit(0.0:13061591): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=18003 scontext=u:r:untrusted_app:s0:c156,c258,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
E/libc    ( 5453): Access denied finding property "vendor.debug.egl.swapinterval"
W/Thread-8( 5453): type=1400 audit(0.0:13061592): avc: denied { read } for name="u:object_r:vendor_default_prop:s0" dev="tmpfs" ino=18003 scontext=u:r:untrusted_app:s0:c156,c258,c512,c768 tcontext=u:object_r:vendor_default_prop:s0 tclass=file permissive=0
E/libc    ( 5453): Access denied finding property "vendor.debug.egl.swapinterval"
```

Is there anyway to stop these? Have I installed it incorrectly.

**Comment by izzytwosheds** [comment]
This looks like how SELinux policy is configured on your device. That is security policy, has nothing to do with LiTr. I don't think we have control over it. I have a feeling this is specific to a device you are running LiTr on, I don't remember seeing these on devices I am testing on (Samsung Galaxy Note 10 and Google Pixel 3)

**Comment by febg11** [comment]
Ah that makes sense... i tried running the project on the emulator and i no longer see all the prints (I guess its just my phone).

However the app seems to crash on the emulator... should litr be able to run on it? I was trying to test everything works on my minSdkVersion which is set to 24 (is this too low for litr?) 

Sorry for more questions its just hard to debug without having a device running api 24.

**Comment by izzytwosheds** [comment]
Always happy to help out. :)
I never tried to run LiTr on emulator, I probably should. I will give it a shot a see if there anything that can be done quickly to get LiTr working on emulator reliably.

**Comment by febg11** [comment]
Thanks i really appreciate it :D

**Comment by izzytwosheds** [comment]
So I did some experimentation. Looks like LiTr works on emulators with API level 26 and above. ExoPlayer (which also relies on MediaCodec) requires emulators with API level 23 and above. I might eventually look into what is happening in versions 23-25, but for now I would recommend testing with 26+ API levels.

**Comment by febg11** [comment]
How have you managed to get it working with api 26? :O 
 
Ive spent the last 2 hours trying to get an mp4 compression working on the emulator with api26... everything I do results in this 

```
D/AudioTrack( 6103): Client defaulted notificationFrames to 6000 for frameCount 12000
E/TransformationJob( 6103): Transformation job error
E/TransformationJob( 6103): com.linkedin.android.litr.exception.TrackTranscoderException: No encoder found.Media transformation failed for job id: null
E/TransformationJob( 6103): Media format: {color-format=2130708361, i-frame-interval=5, durationUs=15933333, mime=video/avc, width=1080, bitrate=1880032, frame-rate=30, height=1920}
E/TransformationJob( 6103): Available media codec info list (Name, IsEncoder, Supported Types): 
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.mp3.decoder,false,[audio/mpeg]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.amrnb.decoder,false,[audio/3gpp]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.amrwb.decoder,false,[audio/amr-wb]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.aac.decoder,false,[audio/mp4a-latm]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.g711.alaw.decoder,false,[audio/g711-alaw]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.g711.mlaw.decoder,false,[audio/g711-mlaw]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.vorbis.decoder,false,[audio/vorbis]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.opus.decoder,false,[audio/opus]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.raw.decoder,false,[audio/raw]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.aac.encoder,true,[audio/mp4a-latm]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.amrnb.encoder,true,[audio/3gpp]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.amrwb.encoder,true,[audio/amr-wb]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.flac.encoder,true,[audio/flac]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.gsm.decoder,false,[audio/gsm]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.mpeg4.decoder,false,[video/mp4v-es]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.h263.decoder,false,[video/3gpp]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.h264.decoder,false,[video/avc]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.hevc.decoder,false,[video/hevc]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.vp8.decoder,false,[video/x-vnd.on2.vp8]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.vp9.decoder,false,[video/x-vnd.on2.vp9]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.h263.encoder,true,[video/3gpp]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.h264.encoder,true,[video/avc]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.mpeg4.encoder,true,[video/mp4v-es]
E/TransformationJob( 6103): MediaCodecInfo: OMX.google.vp8.encoder,true,[video/x-vnd.on2.vp8]
E/TransformationJob( 6103): 	at com.linkedin.android.litr.codec.MediaCodecEncoder.init(MediaCodecEncoder.java:57)
E/TransformationJob( 6103): 	at com.linkedin.android.litr.transcoder.VideoTrackTranscoder.initCodecs(VideoTrackTranscoder.java:74)
E/TransformationJob( 6103): 	at com.linkedin.android.litr.transcoder.VideoTrackTranscoder.<init>(VideoTrackTranscoder.java:64)
E/TransformationJob( 6103): 	at com.linkedin.android.litr.transcoder.TrackTranscoderFactory.create(TrackTranscoderFactory.java:78)
E/TransformationJob( 6103): 	at com.linkedin.android.litr.TransformationJob.createTrackTranscoders(TransformationJob.java:171)
E/TransformationJob( 6103): 	at com.linkedin.android.litr.TransformationJob.transform(TransformationJob.java:94)
E/TransformationJob( 6103): 	at com.linkedin.android.litr.TransformationJob.run(TransformationJob.java:74)
E/TransformationJob( 6103): 	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:457)
E/TransformationJob( 6103): 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
E/TransformationJob( 6103): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1162)
E/TransformationJob( 6103): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:636)
E/TransformationJob( 6103): 	at java.lang.Thread.run(Thread.java:764)
```

I am not sure what the actual error is telling me... it mentions that the job ID is null but Ive definitely checked and it has a unique value. I  cant see how the api 26 would not have the MediaFormat.MIMETYPE_VIDEO_AVC encoder... isn't this included on most devices? 

My setup is this 
```
                        MediaFormat targetVideoFormat = new MediaFormat();
                        targetVideoFormat.setString(MediaFormat.KEY_MIME, MediaFormat.MIMETYPE_VIDEO_AVC);
                        targetVideoFormat.setInteger(MediaFormat.KEY_WIDTH, widthHeightMap.get("width"));
                        targetVideoFormat.setInteger(MediaFormat.KEY_HEIGHT, widthHeightMap.get("height"));
                        targetVideoFormat.setInteger(MediaFormat.KEY_BIT_RATE, videoBitrate);
                        targetVideoFormat.setInteger(MediaFormat.KEY_I_FRAME_INTERVAL, 5);
                        targetVideoFormat.setInteger(MediaFormat.KEY_FRAME_RATE, 30);
```

```
                        MediaFormat audioVideoFormat = new MediaFormat();
                        audioVideoFormat.setString(MediaFormat.KEY_MIME, MediaFormat.MIMETYPE_AUDIO_AAC);
                        audioVideoFormat.setInteger(MediaFormat.KEY_BIT_RATE, audioBitrate);
                        audioVideoFormat.setInteger(MediaFormat.KEY_SAMPLE_RATE, sampleRate);
                        audioVideoFormat.setInteger(MediaFormat.KEY_CHANNEL_COUNT, channel);
```
Which works perfectly on the emulator with api 28.

If a missing encoder is the problem, I have seen that MediaCodecList has a method called ```findEncoderForFormat(MediaFormat format)```. Is this something that can be used to select an encoder to prevent errors on devices  that might be missing encoders?

thank you

**Comment by izzytwosheds** [comment]
I am using a LiTr demo app (with no modifications) and Pixel 3 XL x86 API 26 emulator. Works fine for me. Did you try the demo app on your emulator?

**Comment by febg11** [comment]
Right I was doing something wrong in the flutter code... sorry about that, It's working now! However I have noticed that any portrait videos will get squashed/warped when compressing using the litr example app and it does the same thing in my app when using api 26. 

e.g. 
![image](https://user-images.githubusercontent.com/50048162/89469090-a901d500-d770-11ea-9786-f3e2434cb17d.png)

It seems to take the original portrait video, rotates it 90 degrees so its landscape and then stretches back to the original portrait size. 
 

When doing the compression using the same video on api 28, it doesn't do this. 

I noticed there is rotate settings in the example app but I cannot change them... Is there a setting a can pass in as a parameter to make sure it uses the orginal values properly on all api versions? or is this an actual bug?
